File: node/be/api/venues/site.js
'use strict';

/**
 * The live site's public API: automationexercise.com/api. No key, read-mostly.
 *
 * A quirk worth stating: this API answers HTTP 200 for almost everything and
 * puts the real outcome in a `responseCode` field in the JSON body -- 200, 400,
 * 405. The assertions read that field, not the HTTP status, because that is the
 * contract the API actually offers.
 *
 * A transport failure is SiteUnreachable (grades Blocked). A shape this adapter
 * did not expect is thrown as is.
 */

const BASE = (process.env.SITE_API_URL || 'https://automationexercise.com/api').replace(/\/+$/, '');
const USER_AGENT = 'qa-automation-ecommerce/1.0 (read-only invariants)';
const TIMEOUT_MS = 25_000;

class SiteUnreachable extends Error {}

async function request(method, path, form) {
  const init = { method, headers: { 'user-agent': USER_AGENT, accept: 'application/json' }, signal: AbortSignal.timeout(TIMEOUT_MS) };
  if (form) { init.headers['content-type'] = 'application/x-www-form-urlencoded'; init.body = new URLSearchParams(form).toString(); }
  let res;
  try { res = await fetch(BASE + path, init); } catch (err) {
    throw new SiteUnreachable('GET ' + path + ' -- ' + (err.cause?.code || err.name || err.message));
  }
  const text = await res.text();
  // 5xx/429/403 and Cloudflare-style challenges are the site being unavailable,
  // not the site being wrong. From some runner IPs the JSON API answers 200 with
  // an HTML challenge page; a non-JSON body is treated as unreachable, so the
  // tier grades Blocked rather than throwing on `undefined.products`.
  if (res.status >= 500 || res.status === 429 || res.status === 403) throw new SiteUnreachable(method + ' ' + path + ' -- HTTP ' + res.status);
  let body; try { body = JSON.parse(text); } catch { body = null; }
  if (body === null || typeof body !== 'object') throw new SiteUnreachable(method + ' ' + path + ' -- expected JSON, got ' + (text.slice(0, 40).replace(/\s+/g, ' ')) + '…');
  return { httpStatus: res.status, body, responseCode: body.responseCode ?? null };
}

const productsList = () => request('GET', '/productsList');
const brandsList = () => request('GET', '/brandsList');
const searchProduct = (term) => request('POST', '/searchProduct', term === undefined ? undefined : { search_product: term });
const verifyLogin = (form) => request('POST', '/verifyLogin', form);
const productsListPost = () => request('POST', '/productsList');   // method not allowed by the API
const brandsListPost = () => request('POST', '/brandsList');
const getUserByEmail = (email) => request('GET', '/getUserDetailByEmail?email=' + encodeURIComponent(email));

/** Parse "Rs. 500" into an integer number of rupees; null if it does not match. */
function parsePrice(s) {
  const m = /^Rs\.\s*(\d+)$/.exec(String(s).trim());
  return m ? Number(m[1]) : null;
}

module.exports = { BASE, SiteUnreachable, request, productsList, brandsList, searchProduct, verifyLogin, productsListPost, brandsListPost, getUserByEmail, parsePrice };

File: python/be/api/venues/site.py
"""The live site's public API: automationexercise.com/api. Mirror of
node/be/api/venues/site.js. Answers HTTP 200 with a responseCode in the body;
the assertions read that field. Transport failure is SiteUnreachable.
"""

from __future__ import annotations

import json
import os
import re
import urllib.error
import urllib.parse
import urllib.request

BASE = os.environ.get("SITE_API_URL", "https://automationexercise.com/api").rstrip("/")
USER_AGENT = "qa-automation-ecommerce/1.0 (read-only invariants)"
TIMEOUT = 25


class SiteUnreachable(Exception):
    pass


def request(method, path, form=None):
    headers = {"user-agent": USER_AGENT, "accept": "application/json"}
    data = None
    if form is not None:
        headers["content-type"] = "application/x-www-form-urlencoded"
        data = urllib.parse.urlencode(form).encode()
    req = urllib.request.Request(BASE + path, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=TIMEOUT) as res:
            status, text = res.status, res.read().decode("utf-8", "replace")
    except urllib.error.HTTPError as err:
        if err.code >= 500 or err.code in (429, 403):
            raise SiteUnreachable(f"{method} {path} -- HTTP {err.code}") from err
        status, text = err.code, err.read().decode("utf-8", "replace")
    except (urllib.error.URLError, TimeoutError, OSError) as err:
        raise SiteUnreachable(f"{method} {path} -- {err}") from err
    # From some runner IPs the JSON API answers 200 with an HTML challenge page;
    # a non-JSON body is the site being unavailable, not wrong -- grade Blocked.
    try:
        body = json.loads(text)
    except json.JSONDecodeError:
        body = None
    if not isinstance(body, dict):
        raise SiteUnreachable(f"{method} {path} -- expected JSON, got " + " ".join(text[:40].split()) + "…")
    return {"httpStatus": status, "body": body, "responseCode": body.get("responseCode")}


def products_list():
    return request("GET", "/productsList")


def brands_list():
    return request("GET", "/brandsList")


def search_product(term):
    return request("POST", "/searchProduct", None if term is None else {"search_product": term})


def verify_login(form):
    return request("POST", "/verifyLogin", form)


def products_list_post():
    return request("POST", "/productsList")


def brands_list_post():
    return request("POST", "/brandsList")


def get_user_by_email(email):
    return request("GET", "/getUserDetailByEmail?email=" + urllib.parse.quote(email))


def parse_price(s):
    m = re.match(r"^Rs\.\s*(\d+)$", str(s).strip())
    return int(m.group(1)) if m else None
